third_party/libtock-drivers/src/nfc.rs

```diff
@@ -0,0 +1,128 @@
+use crate::result::TockResult;
+use crate::util;
+use core::cell::Cell;
+use core::mem;
+use libtock_core::{callback, syscalls};
+
+const DRIVER_NUMBER: usize = 0x30003;
+
+mod command_nr {
+    pub const CHECK: usize = 0;
+    pub const TRANSMIT: usize = 1;
+    pub const RECEIVE: usize = 2;
+    pub const EMULATE: usize = 3;
+    pub const CONFIGURE: usize = 4;
+    pub const FRAMEDELAYMAX: usize = 5;
+}
+
+mod subscribe_nr {
+    pub const TRANSMIT: usize = 1;
+    pub const RECEIVE: usize = 2;
+    pub const SELECT: usize = 3;
+}
+
+mod allow_nr {
+    pub const TRANSMIT: usize = 1;
+    pub const RECEIVE: usize = 2;
+}
+
+#[allow(dead_code)]
+pub struct RecvOp {
+    pub result_code: usize,
+    pub recv_amount: usize,
+}
+
+pub struct NfcTag {}
+
+impl NfcTag {
+    /// Check the existence of an NFC driver.
+    pub fn setup() -> bool {
+        syscalls::command(DRIVER_NUMBER, command_nr::CHECK, 0, 0).is_ok()
+    }
+
+    pub fn enable_emulation() -> bool {
+        NfcTag::emulate(true)
+    }
+
+    pub fn disable_emulation() -> bool {
+        NfcTag::emulate(false)
+    }
+
+    fn emulate(enabled: bool) -> bool {
+        syscalls::command(DRIVER_NUMBER, command_nr::EMULATE, enabled as usize, 0).is_ok()
+    }
+
+    /// Subscribe to the tag being SELECTED callback.
+    pub fn selected() -> bool {
+        let is_selected = Cell::new(false);
+        let mut is_selected_alarm = || is_selected.set(true);
+        let subscription = syscalls::subscribe::<callback::Identity0Consumer, _>(
```

Comment:
I'm wondering about the sequence of events here. We subscribe to a callback but without triggering anything with a command, which is unusual with Tock.

How does the driver get notified that it should trigger the callback? Is it triggered just by the subscription (which although unusual is fine)?

I just want to make sure that the application doesn't miss any callback that the kernel would trigger before any subscription has happened.

Reply:
So this is actually an issue that I didn't consider. I can't think of a solution other than adding a command to unmask the `Selected` interrupt.